python_modules/dagster/dagster_tests/asset_defs_tests/test_assets.py

```diff
@@ -156,6 +156,40 @@ def abc():
     }
 
 
+def test_with_replaced_metadata() -> None:
+    @multi_asset(
+        outs={
+            "foo": AssetOut(metadata={"foo": "foo"}),
+            "bar": AssetOut(metadata={"bar": "bar"}),
+            "baz": AssetOut(metadata={"baz": "baz"}),
+        }
+    )
+    def abc():
+        ...
+
+    assert abc.metadata_by_key == {
+        AssetKey("foo"): {"foo": "foo"},
+        AssetKey("bar"): {"bar": "bar"},
+        AssetKey("baz"): {"baz": "baz"},
+    }
+
+    # If there's no replacement metadata for the asset key, the original metadata is retained.
```

Comment:
This logic differs from (e.g.) descriptions_by_key, which on closer inspection seems to wipe out existing descriptions when this method is invoked without a value there. I also think we should distinguish between an empty dictionary and `None`, as passing in an empty dictionary seems like "there is no metadata" rather than "just keep whatever metadata you have". High level I think we should do something like:

```
thing_by_key = check.dict_param(...) if thing_by_key is not None else self.thing_by_key

replaced_thing_by_key = {output_key_replacements.get(key, key): value for key, value in thing_by_key.items()}
```

Reply:
This is actually the same behavior as `descriptions_by_key`.

https://github.com/dagster-io/dagster/blob/d876bde018919fbbb252f8078a6a57627d22b2d8/python_modules/dagster/dagster_tests/asset_defs_tests/test_assets.py#L142-L156